File: leetcode/exercises/1.5.1.medium.top_k_frequent_elements_2.js
// -----------------------
// Created       : 27/10/2025
// Last Edited   : 27/10/2025
// Big O         :
// Topics        : 
// Problem Id    : 347
// -----------------------

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    const counts = {}
    for (num of nums)
    {
        counts[num] = (counts[num] ?? 0) + 1;
    };

    const frequency = Array.from({length: nums.length}, () => []);
    for (num of Object.keys(counts))
    {
        frequency[counts[num]-1].push(Number(num));
    };

    res = []
    for (let i = frequency.length-1; i >= 0; i--)
    {
        for (const j of frequency[i])
        {
            if (res.length !== k){
                res.push(j);
            };
        };
    };

    return res;
};

console.log(topKFrequent([1,1,2,2,3,3,3,5,6,5], 3));

File: leetcode/exercises/1.5.1.medium.top_k_frequent_elements_2.py
# -----------------------
# Created     : 21/10/2025
# Last Edited : 24/10/2025
# Topics      : Bucket Sort
# Big O       :
# Problem Id  : 347
# Source      :
# Notes       :
# -----------------------

class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        if len(nums) == 0:
            return []

        counts = {}
        for num in nums:
            counts[num] = counts.get(num, -1) + 1

        freq = [[] for i in range(len(nums))]
        for num, count in counts.items():
            freq[count].append(num)

        result = []
        for i in range(len(freq)-1, -1, -1):
            if len(freq[i]) == 0:
                continue

            for j in freq[i]:
                result.append(j)
                if len(result) == k:
                    return result

sol = Solution()
print()
print(sol.topKFrequent([11, 1, 2, 2, 2, 3, 4, 5, 5, 5, 5, 6], 3))
